docs/DMS/DOCUMENT-AUTHORIZATION.md

```diff
@@ -0,0 +1,268 @@
+document (e.g. SchoolId, EducationServiceCenterId, Namespace) with a list of authorizations configured for
+each client. Additionally, EducationOrganizations form a hierarchy (e.g. Schools are a part of a
+LocalEducationAgency) where a higher-level client authorization applies to lower level security attributes.
+
+EducationOrganization-based security also extends the authorization hierarchy down to the Student level. This
+flows through Associations such as StudentSchoolAssociation to define the student data that is authorized for
+a client, based on their EducationOrganization permissions.
+
+The design of Client Authorizations is covered in [DMS-AUTH](DMS-AUTH.md).
+
+The scope of the problem is to determine how the DMS will:
+
+- Efficiently extract the security attributes of a document.
+- Understand the hierarchy of EducationOrganizations.
+- Understand the extension of the EducationOrganization hierarchy down to the Student level.
+- Efficiently apply a client's authorizations to POST, PUT, GET and DELETE actions, including to a search
+  engine datastore.
+
+## Solution
+
+### Security Attribute Location
+
+DMS needs to know which resources have security attributes and where those attributes are located on the
+document. These attributes then need to be extracted into a `securityAttributes` object.
+
+MetaEd will provide this information in the ResourceSchema in ApiSchema.json for any EducationOrganization,
+Student, and Namespace field on a resource. There will be an entry for each type of security attribute for the
+resource along with the attribute JsonPath. This is independent of run-time security configuration to allow
+for support of all configurations. For example, Course has an EducationOrganization superclass reference and
+so the SecurityAttribute would look like this:
+
+```json
+"courses": {
+  "securityAttributes": {
+    "EducationOrganization": [
+      "$.educationOrganizationReference.educationOrganizationId"
+    ]
+  }
+}
+```
+
+DisciplineAction provides an example of a School (an EducationOrganization subclass) reference with a role
+name:
+
+```json
+"disciplineActions": {
+  "securityAttributes": {
+    "School": [
+      "$.responsibilitySchoolReference.schoolId"
+    ]
+  }
+}
+```
+
+Similarly, Credentials has a Namespace and would look like:
+
+```json
+"credentials": {
+  "securityAttributes": {
+    "Namespace": [
+      "$.namespace"
+    ]
+  }
+}
+```
+
+And finally, StudentGradebookEntry has a Student and would look like:
+
+```json
+"credentials": {
+  "securityAttributes": {
+    "StudentUniqueId": [
+      "$.studentReference.studentUniqueId"
+    ]
+  }
+}
+```
+
+The `securityAttributes` object can have multiple attributes (for example EducationOrganization, School, and
+StudentUniqueId) - and each entry is an array to allow for multiple JsonPaths.
+
+Note that this scheme requires MetaEd to have hardcoded knowledge of Namespace, StudentUniqueId and the
```

Comment:
Yes, this would be at a MetaEd enhancer level. It seems natural to tag them ahead of time rather that figure it out at runtime with DMS. Without some sort of hinting in ApiSchema.json, I think DMS would have to rely on naming to figure it out - which is very brittle - so might as well just spell it out and keep DMS simple. Keeping DMS simple whenever possible is also friendly to alternative Ed-Fi API implementations that want to take advantage of ApiSchema.json